theano/sandbox/cuda/cusolver.py

```diff
@@ -0,0 +1,221 @@
+import ctypes
+import theano
+from theano.sandbox.cuda import (CudaNdarrayType, as_cuda_ndarray_variable,
+                                 GpuOp, cuda_ndarray, CudaNdarray)
+from theano.tensor import reshape
+try:
+    from scikits.cuda.cusolver import (_libcusolver, cusolverDnCreate,
+                                       cusolverCheckStatus)
+    from scikits.cuda.cublas import _CUBLAS_SIDE_MODE, _CUBLAS_OP
+    cusolver_available = True
+except (ImportError, OSError):
+    cusolver_available = False
+    raise
+
+dimshuffle = cuda_ndarray.cuda_ndarray.dimshuffle
+
+
+class CuSolverOp(GpuOp):
+    """cuSOLVER ops lazily raise an error when cuSOLVER is not available."""
+    def make_thunk(self, *args, **kwargs):
+        if not cusolver_available:
+            raise RuntimeError('cuSOLVER not available')
+        return self._make_thunk(*args, **kwargs)
+
+
+class GpuGeqrf(CuSolverOp):
+    """Wrapper of cuSOLVER's implementation of the geqrf LAPACK method.
+
+    Notes
+    -----
+    Wrapping of the methods is done as in `scikits.cuda.cusolver`. The
+    interface is intendend to mimick `scipy.linalg.lapack`.
+
+    """
+    __props__ = ()
+
+    def output_type(self):
+        return [CudaNdarrayType(broadcastable=[False, False])(),
+                CudaNdarrayType(broadcastable=[False])(),
+                CudaNdarrayType(broadcastable=[False])(),
+                CudaNdarrayType(broadcastable=[])()]
+
+    def make_node(self, inp):
+        inp = as_cuda_ndarray_variable(inp)
+        assert inp.ndim == 2
+
+        return theano.Apply(self, [inp], self.output_type())
+
+    def _make_thunk(self, node, storage_map, _, no_recycling=[]):
+        inputs = [storage_map[v] for v in node.inputs]
+        outputs = [storage_map[v] for v in node.outputs]
+
+        def thunk():
+            A = inputs[0][0]
+            m, n = A.shape
+            # LAPACK routines expect Fortran order
+            A_T = dimshuffle(A, (1, 0)).reshape((m, n))
+            A_copy = A_T.copy()
+
+            lda = max(1, m)
+            TAU = CudaNdarray.zeros((min(m, n),))
+            Lwork = cusolverDnSgeqrf_bufferSize(handle, m, n, A_copy.gpudata,
+                                                lda)
+            Work = CudaNdarray.zeros((max(1, Lwork),))
```

Comment:
If these allocations become too inefficient, we could add these buffers as additional outputs. That way, if the garbage collector is disabled, the Op could try to re-use them.